crates/wasi-common/src/sys/unix/oshandle.rs

```diff
@@ -1,123 +1,59 @@
-use crate::entry::EntryRights;
-use crate::sys::oshandle::{AsFile, OsHandle, OsHandleExt};
-use crate::wasi::{types, RightsExt};
-use std::fs::{File, OpenOptions};
+use crate::sys::AsFile;
+use std::cell::Cell;
+use std::fs::File;
 use std::io;
 use std::mem::ManuallyDrop;
-use std::os::unix::prelude::{AsRawFd, FileTypeExt, FromRawFd, IntoRawFd, RawFd};
+use std::os::unix::prelude::{AsRawFd, FromRawFd, IntoRawFd, RawFd};
 
-pub(crate) use super::sys_impl::osfile::*;
+#[derive(Debug)]
+pub(crate) struct OsHandle(Cell<RawFd>);
 
-impl AsRawFd for OsHandle {
-    fn as_raw_fd(&self) -> RawFd {
-        match self {
-            Self::OsFile(file) => file.as_raw_fd(),
-            Self::Stdin => io::stdin().as_raw_fd(),
-            Self::Stdout => io::stdout().as_raw_fd(),
-            Self::Stderr => io::stderr().as_raw_fd(),
+impl OsHandle {
+    /// Tries clone `self`.
+    pub(crate) fn try_clone(&self) -> io::Result<Self> {
+        let fd = self.as_file()?.try_clone()?;
+        Ok(Self(Cell::new(fd.into_raw_fd())))
+    }
+    /// Consumes `other` taking the ownership of the underlying
+    /// `RawFd` file descriptor.
+    ///
+    /// Note that the state of `Dir` stream pointer *will* not be carried
+    /// across from `other` to `self`.
+    pub(crate) fn update_from(&self, other: Self) {
+        let new_fd = other.into_raw_fd();
+        let old_fd = self.0.get();
+        self.0.set(new_fd);
+        // We need to remember to close the old_fd.
+        unsafe {
+            File::from_raw_fd(old_fd);
         }
     }
```

Comment:
If I understand correctly, the main use for `update_from` is for the `fdstat_set_flags`, which on Windows opens a new handle. If that's the only thing that needs it, could we make `update_from` on Unix just panic?

And, if we don't need to implement `update_from`, can `OsHandle` on Unix hold a plain `RawFd` instead of a `Cell<RawFd>`?

Reply:
Makes sense!

Reply:
OK, it's done now. Since, as you rightly pointed out in one of the comments above, we don't really need interior mutability on *nix wrt the inner `RawFd` instance, I've also replaced `RawFd` with `File` instance which takes care of a couple of ownership related things for us. I believe this is what @iximeow was suggesting all along as well. Now that @sunfishcode pointed out that issuing `RawOsHandle::update_from` call on *nix is a bug, using `File` was now possible.